src/Integration/Binding/SolutionBindingOperation.cs

```diff
@@ -0,0 +1,204 @@
+        {
+        }
+
+        internal SolutionBindingOperation(IServiceProvider serviceProvider, IProjectSystemHelper projectSystemHelper, string sonarQubeProjectKey, SolutionRuleSetWriter solutionRuleSetWriter)
+        {
+            if (serviceProvider == null)
+            {
+                throw new ArgumentNullException(nameof(serviceProvider));
+            }
+
+            if (projectSystemHelper == null)
+            {
+                throw new ArgumentNullException(nameof(projectSystemHelper));
+            }
+
+            if (string.IsNullOrWhiteSpace(sonarQubeProjectKey))
+            {
+                throw new ArgumentNullException(nameof(sonarQubeProjectKey));
+            }
+
+            this.serviceProvider = serviceProvider;
+            this.projectSystem = projectSystemHelper;
+            this.solutionRuleSetWriter = solutionRuleSetWriter ?? new SolutionRuleSetWriter(sonarQubeProjectKey);
+
+        }
+
+        #region State
+        internal /*for testing purposes*/ IList<IBindingOperation> Binders
+        {
+            get { return this.childBinder; }
+        }
+
+        internal /*for testing purposes*/ string SolutionFullPath
+        {
+            get;
+            private set;
+        }
+
+        internal /*for testing purposes*/ IReadOnlyDictionary<RuleSetGroup, RuleSetInformation> RuleSetsInformationMap
+        {
+            get { return this.ruleSetsInformationMap; }
+        }
+        #endregion
+
+        #region ISolutionRuleStore
+
+        public void RegisterKnownRuleSets(IDictionary<RuleSetGroup, RuleSet> ruleSets)
+        {
+            if (ruleSets == null)
+            {
+                throw new ArgumentNullException(nameof(ruleSets));
+            }
+
+            foreach (var keyValue in ruleSets)
+            {
+                Debug.Assert(!this.ruleSetsInformationMap.ContainsKey(keyValue.Key), "Attempted to register an already registered rule set. Group:" + keyValue.Key);
+                this.ruleSetsInformationMap[keyValue.Key] = new RuleSetInformation(keyValue.Key, keyValue.Value);
+            }
+        }
+
+        public string GetRuleSetFilePath(RuleSetGroup group)
+        {
+            RuleSetInformation info;
+
+            if (!this.ruleSetsInformationMap.TryGetValue(group, out info) || info == null)
+            {
+                Debug.Fail("Expected to be called by the ProjectBinder after the known rulesets were registered");
+                return null;
+            }
+
+            Debug.Assert(info.NewRuleSetFilePath != null, "Expected to be called after Prepare");
+
+            // At this point we know that the rule set is needed, so we write it to disk (internally will only write the file once)
+            return info.NewRuleSetFilePath;
+        }
+
+        #endregion
+
+        #region IBindingOperation
+        public void Initialize()
+        {
+            this.SolutionFullPath = this.projectSystem.GetCurrentActiveSolution().FullName;

```

Comment:
Why? What will it solve? Is there an invariant that we can really check here?
The only case when we want is when the user closed it, in which case the workflow will be aborted but an exception will still be raised and logged in the output. The reason is because there's an internal between when the user clicks bind and the code actually reaches here, it should be fairly short in practice for the binding workflow not to be cancelled before hitting this code (this run on the UI thread). Don't think we need to do anything special here. Let's discuss offline if you feel differently.